enhancements/olm/platform-operators.md

```diff
@@ -0,0 +1,435 @@
+CVO->>ClusterOperator: Watches
+```
+
+Note: The above workflow requires admins to manually create, and edit, the cluster Kubernetes manifests with their desired PO intent. There's a chance that admins mistype the package names, and depending on whether upfront validation is possible, that invalid configuration may not be surfaced until after cluster creation. This behavior may require explicit documentation to reduce this risk. These are obvious UX gaps with this workflow, which can be solidified in future phases, but it will result in zero installer changes during the 4.12 timeframe.
+
+### API Extensions
+
+- Adds a top-level PlatformOperator cluster-scoped CRD.
+- Adds the Bundle and BundleDeployment OLM 1.x API(s).
+- Adds a validating webhook for the Bundle and BundleDeployment APIs.
+- Adds a CRD validation webhook.
+
+### Risks and Mitigations
+
+- Early phases will introduce multiple v1alpha1 CRDs downstream.
+- Early phases will introduce OLM v1 components that are in alpha stages.
+- No concrete migration plan from legacy to OLM v1.
+- Coordination around external team dependencies.
+- The BundleDeployment API has a large net of permissions that may require upfront RBAC to limit this API to admins.
+
+### Drawbacks
+
+- Implementing this PO mechanism using legacy OLM would provide more immediate business value. That implementation is outlined below in the alternatives section.
+- Early phases will deliver through tech preview guidelines. This leads to little customer feedback, given it precludes production clusters.
+- Early phases may introduce a new OLM 1.x control plane that has little-to-no interaction with the existing, legacy OLM 0.x control plane.
+- Early phases may introduce validating webhooks. These webhooks can impact the cluster's overall health when they're not available.
+- Early phases may introduce validating webhooks. By default, these webhooks will be deployed using an HA configuration, but may not have logic to accommodate SNO cluster environments.
+- Early phases may introduce a dynamic watching implementation for managing the underlying OLM operator resources. This has potential performance implications that need to be measured through benchmarking and/or addressed before any GA phase.
+- Future phase designs may require a large scale rewrite of the OLM component.
+
+## Design Details
+
+### Open Questions [optional]
+
+- Does the console need to be able to provide a UI experience for POs?
+- Does having separate APIs for ClusterOperators and PlatformOperators provide a degraded admin UX?
+- How to handle the maxOpenShiftVersion constraint given it's defined in the CSV resource?
+- Dedicated CatalogSource? Dedicated channel? The top-level PO API contains a catalog configuration?
+- Is it possible to have a dedicated PO index image?
+- Should the existing minKubeVersion and maxOpenShiftVersion fields be unified?
+- How are phase 0 requirements enforced?
+- Does this phase need to enforce cluster invariants?
```

Comment:
such as?

Reply:
OLM today enforces two cluster invariants:

- No two operators may provide the same package in the same namespace
- No two operators may provide the same GVK in the same namespace

For phase 1+, this may mean enforcing that all POs are sourced from the same 4.x channel, e.g. "4.12", if we're interested in providing safeguards for what PO content can be installed in subsequent phases. Enforcing cluster invariants will likely need to wait until phase 1, or whenever deppy comes into play.

Reply:
i don't think we need to worry about enforcing those invariants in p0.  Given the limited use it's going to get, i think "don't do things like install the same package twice or install two things that provide the same GVK" can be documented

Reply:
> Is there a reason to be flexible with the namespace that a platform operator can run in?

some amount of flexibility, yes.  Likely admins will want the operators running in unique NSes in most cases (but perhaps not all) and we can't let operators hardcode an NS for fear we end up w/ collisions.  

but this is more about what NSes the operator is watching, for which we definitely need constraints to avoid two operators trying to reconcile the same resource objects in a given NS.

Reply:
> Maybe this behavior is something to revisit during subsequent phases?

it will definitely be important as we widen the discussion to include more direct language about core cluster operators, or platform specific operators that supplement our core operators. afaik no one has created this type of core operator yet that includes the platform specific behavior, but i think it will be more important as we start to tell users how to do these types of transformations.

as long as we have a note to revisit this topic when we push into the phases that speak directly to the core/platform operators, then i'm ok. can we document this as follow up work?

Reply:
I think we captured $some of this already as non-goals + open questions. I don't think there's any language around speaking direction to core/platform operators right now, but I'd imagine that could be tracked as a spike ticket. The current plan is to handle a small pool of OLM-based operators, then lessen the restrictions next phase, then start working towards onboarding current CVO-based operators. And we could re-evaluate this direction for that last direction.